Compiler/src/verifytrim.jl

```diff
@@ -49,76 +49,223 @@ const runtime_functions = Symbol[
-        return newstmt
-    elseif newstmt isa Callable
-        return newstmt
     end
-    return Expr(:(::), newstmt, typ)
+    return true
+end
+
+# Check if a type is "stable" - uses same classification as code_warntype
+function is_type_stable(@nospecialize(typ))
+    typ isa Core.Const && return true
+    typ === Union{} && return false
+    typ == Core.Box && return false
+    typ isa Union && return is_expected_union(typ)
+    return isdispatchelem(typ)
+end
+
+# Color printing for types - same colors as code_warntype:
+# - red bold: unstable/abstract types
+# - yellow: expected unions (small unions of concrete types)
+# - light_black: stable concrete types
+function print_type_colored(io::IO, @nospecialize(typ))
+    color = get(io, :color, false)::Bool
+    if !color
+        print(io, typ)
+    elseif typ === Union{}
+        printstyled(io, typ; color=:red)
+    elseif typ == Core.Box
+        printstyled(io, typ; color=:red, bold=true)
+    elseif typ isa Union && is_expected_union(typ)
+        printstyled(io, typ; color=:yellow)
+    elseif isdispatchelem(typ)
+        printstyled(io, typ; color=:light_black)
+    else
+        printstyled(io, typ; color=:red, bold=true)
+    end
+end
+
+# Print with light_black if stable, normal otherwise
+function print_styled(io::IO, @nospecialize(x), stable::Bool)
+    stable ? printstyled(io, x; color=:light_black) : print(io, x)
+end
+
+# Check if we should elide the type annotation for a statement
+function should_elide_type(newstmt, @nospecialize(typ))
+    newstmt isa Expr && newstmt.head ∈ (:quote, :inert) && return true
+    newstmt isa GlobalRef && isdispatchelem(typ) && return true
+    newstmt isa Union{Int, UInt8, UInt16, UInt32, UInt64, Float16, Float32, Float64, String, QuoteNode} && return true
+    newstmt isa Callable && return true
+    return false
 end
 
-# map the ssavalues in a (value-producing) statement to the expression they came from, summarizing some things to avoid excess printing
-function mapssavalues(codeinfo::CodeInfo, sptypes::Vector{VarState}, stmt)
-    @nospecialize stmt
-    if stmt isa SSAValue
-        return mapssavalues(codeinfo, sptypes, codeinfo.code[stmt.id])
-    elseif stmt isa PiNode
-        return mapssavalues(codeinfo, sptypes, stmt.val)
+# Unwrap SSAValue and PiNode to get the underlying statement
+function unwrap_stmt(codeinfo::CodeInfo, @nospecialize(stmt))
+    while true
+        if stmt isa SSAValue
+            stmt = codeinfo.code[stmt.id]
+            isexpr(stmt, :(=)) && (stmt = stmt.args[2])
+        elseif stmt isa PiNode
+            stmt = stmt.val
+        else
+            return stmt
+        end
+    end
+end
+
+# Convert Core.Argument to slot name symbol if available
+function argument_name(codeinfo::CodeInfo, arg::Core.Argument)
+    slotnames = codeinfo.slotnames
+    if slotnames !== nothing && arg.n <= length(slotnames)
```

Comment:
The IR verifier seems to suggest that `codeinfo.slotnames` is always populated and the right length - Is that not true?